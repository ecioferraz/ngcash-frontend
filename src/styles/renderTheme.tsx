import { render, RenderResult } from '@testing-library/react';
import { ReactNode } from 'react';
import { ThemeProvider } from 'styled-components';
import { theme } from './theme';

export const renderTheme = (children: ReactNode): RenderResult => {
  return render(<ThemeProvider theme={theme}>{children}</ThemeProvider>);
};
