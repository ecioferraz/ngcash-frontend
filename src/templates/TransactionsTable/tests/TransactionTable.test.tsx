import renderTheme from '../../../styles/renderTheme';
import TransactionTable from '..';
import { transactionsMock } from './mocks';
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import AccountProvider from '../../../providers/AccountProvider';

describe('<TransactionTable />', () => {
  describe('with transactions history', () => {
    let container: HTMLElement;

    beforeEach(() => {
      container = renderTheme(
        <AccountProvider>
          <TransactionTable transactions={transactionsMock} />,
        </AccountProvider>,
      ).container;
    });

    it('should render <TransactionTable /> with transactions history', () => {
      expect(container).toMatchSnapshot();
    });

    it('should select a different type of transaction', async () => {
      const select = screen.getByRole('combobox');
      const all = screen.getByRole('option', { name: 'Todas' });
      const cashIn = screen.getByRole('option', { name: 'Entradas' });

      expect((all as HTMLOptionElement).selected).toBe(true);

      await userEvent.selectOptions(select, cashIn);

      expect((cashIn as HTMLOptionElement).selected).toBe(true);
    });

    it('should render a different svg button when the order of transactions gets changed', async () => {
      let dropDown = container.querySelector('svg[name="desc"]');

      expect(dropDown).toBeInTheDocument();

      await userEvent.click(dropDown?.parentElement as HTMLButtonElement);

      const dropUp = container.querySelector('svg[name="asc"]');

      expect(dropUp).toBeInTheDocument();

      await userEvent.click(dropUp?.parentElement as HTMLButtonElement);

      dropDown = container.querySelector('svg[name="desc"]');

      expect(dropDown).toBeInTheDocument();
    });

    it("should render 'Mostrar menos' button", async () => {
      const showMoreBtn = screen.getByRole('button', { name: 'Mostrar mais' });

      expect(showMoreBtn).toBeInTheDocument();

      await userEvent.click(showMoreBtn);

      const showLessBtn = screen.getByRole('button', { name: 'Mostrar menos' });

      expect(showLessBtn).toBeInTheDocument();
    });
  });

  describe('without transactions history', () => {
    it('should render <Loading />', () => {
      const { container } = renderTheme(<TransactionTable loading />);

      expect(container).toMatchSnapshot();
    });

    it('should render a message when there are no transactions to show', () => {
      renderTheme(<TransactionTable />);

      expect(
        screen.getByText('Você ainda não realizou transações.'),
      ).toBeInTheDocument();
    });

    it('should render <TransactionTable /> with an error message', () => {
      const { container } = renderTheme(
        <TransactionTable error="Unauthorized" />,
      );

      expect(container).toMatchInlineSnapshot(`
        .c1 {
          font-size: 3.2rem;
          text-transform: uppercase;
        }

        .c2 {
          font-size: 1.6rem;
          text-transform: none;
        }

        .c0 {
          text-align: center;
          width: 100%;
        }

        .c0 > * {
          margin: 2rem 0;
        }

        .c0 > h1 {
          margin-bottom: 0;
          margin-top: 1rem;
        }

        .c0 > button {
          margin-top: 0;
        }

        @media (max-width:768px) {
          .c0 {
            width: -webkit-fit-content;
            width: -moz-fit-content;
            width: fit-content;
          }
        }

        <div>
          <section
            class="c0"
          >
            <h1
              class="c1"
            >
              Histórico de transações
            </h1>
            <p
              class="c2"
            >
              Unauthorized
            </p>
          </section>
        </div>
      `);
    });
  });
});
