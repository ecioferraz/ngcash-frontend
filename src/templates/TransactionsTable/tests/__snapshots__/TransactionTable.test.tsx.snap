// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TransactionTable /> with transactions history should render <TransactionTable /> with transactions history 1`] = `
.c7 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c6 {
  background: #000000;
  border: 1px solid #000000;
  border-radius: 4px;
  box-shadow: #DDDDDD 3px 3px 0 0,#000000 3px 3px 0 1px;
  box-sizing: border-box;
  color: #DDDDDD;
  cursor: pointer;
  display: inline-block;
  line-height: 20px;
  overflow: visible;
  padding: 6px 24px;
  touch-action: manipulation;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-user-select: none;
  vertical-align: middle;
  white-space: nowrap;
}

.c6:disabled {
  pointer-events: none;
}

.c6:focus {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c6:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c6:not([disabled]):active {
  background: #DDDDDD;
  outline: 0;
  -webkit-transition: none;
  transition: none;
  box-shadow: #DDDDDD 2px 2px 0 0,#000000 2px 2px 0 1px;
  -webkit-translate: 2px,2px;
  translate: 2px,2px;
}

.c6:has(svg) {
  padding: 0.5% 0.7%;
}

.c6 > svg {
  height: 1.6rem;
  width: 1.6rem;
}

.c5 {
  background: transparent;
  border: none;
  border-bottom: solid 1px #DDDDDD;
  color: #FFFFFF;
  font-size: 1.6rem;
  outline: none;
  width: 10rem;
}

.c5:focus {
  border: none;
  border-bottom: 1px solid #00D775;
  outline: none;
}

.c5 > option {
  background: #000000;
  font-size: 1.6rem;
  height: 0.8rem;
}

.c8 {
  border: 1px solid #000000;
  border-radius: 4px;
  box-shadow: #DDDDDD 3px 3px 0 0,#000000 3px 3px 0 1px;
  box-sizing: border-box;
  border-collapse: collapse;
  font-size: 1.6rem;
  width: 100%;
}

.c8 * {
  padding: 1rem;
  border: 1px solid #FFFFFF;
  text-align: center;
}

.c1 {
  font-size: 3.2rem;
  text-transform: uppercase;
}

.c4 {
  font-size: 1.6rem;
  text-transform: none;
}

.c0 {
  text-align: center;
  width: 100%;
}

.c0 > * {
  margin: 2rem 0;
}

.c0 > h1 {
  margin-bottom: 0;
  margin-top: 1rem;
}

.c0 > button {
  margin-top: 0;
}

.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
}

.c3 > p {
  margin: 0;
}

@media (max-width:768px) {
  .c0 {
    width: -webkit-fit-content;
    width: -moz-fit-content;
    width: fit-content;
  }
}

<div>
  <section
    class="c0"
  >
    <h1
      class="c1"
    >
      Histórico de transações
    </h1>
    <section
      class="c2"
    >
      <section
        class="c3"
      >
        <p
          class="c4"
        >
          Tipo:
        </p>
        <select
          class="c5"
        >
          <option
            value="all"
          >
            Todas
          </option>
          <option
            value="cash-in"
          >
            Entradas
          </option>
          <option
            value="cash-out"
          >
            Saídas
          </option>
        </select>
      </section>
      <section
        class="c3"
      >
        <p
          class="c4"
        >
          Ordenar por data:
        </p>
        <button
          class="c6"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="c7"
            fill="currentColor"
            focusable="false"
            name="drop-down"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m7 10 5 5 5-5H7z"
            />
          </svg>
        </button>
      </section>
    </section>
    <table
      class="c8"
    >
      <thead>
        <tr>
          <th>
            De/para
          </th>
          <th>
            Valor (R$)
          </th>
          <th>
            Tipo
          </th>
          <th>
            Data
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            username1
          </td>
          <td>
            5.00
          </td>
          <td>
            cash-in
          </td>
          <td>
            01/01/2023 15:57:21
          </td>
        </tr>
        <tr>
          <td>
            username2
          </td>
          <td>
            5.00
          </td>
          <td>
            cash-out
          </td>
          <td>
            02/01/2023 11:20:34
          </td>
        </tr>
        <tr>
          <td>
            username1
          </td>
          <td>
            5.00
          </td>
          <td>
            cash-in
          </td>
          <td>
            03/01/2023 15:57:21
          </td>
        </tr>
        <tr>
          <td>
            username2
          </td>
          <td>
            5.00
          </td>
          <td>
            cash-out
          </td>
          <td>
            04/01/2023 11:20:34
          </td>
        </tr>
        <tr>
          <td>
            username1
          </td>
          <td>
            5.00
          </td>
          <td>
            cash-in
          </td>
          <td>
            05/01/2023 15:57:21
          </td>
        </tr>
        <tr>
          <td>
            username2
          </td>
          <td>
            5.00
          </td>
          <td>
            cash-out
          </td>
          <td>
            06/01/2023 11:20:34
          </td>
        </tr>
        <tr>
          <td>
            username1
          </td>
          <td>
            5.00
          </td>
          <td>
            cash-in
          </td>
          <td>
            07/01/2023 15:57:21
          </td>
        </tr>
        <tr>
          <td>
            username2
          </td>
          <td>
            5.00
          </td>
          <td>
            cash-out
          </td>
          <td>
            08/01/2023 11:20:34
          </td>
        </tr>
        <tr>
          <td>
            username1
          </td>
          <td>
            5.00
          </td>
          <td>
            cash-in
          </td>
          <td>
            09/01/2023 15:57:21
          </td>
        </tr>
        <tr>
          <td>
            username2
          </td>
          <td>
            5.00
          </td>
          <td>
            cash-out
          </td>
          <td>
            10/01/2023 11:20:34
          </td>
        </tr>
      </tbody>
    </table>
    <button
      class="c6"
      type="button"
    >
      Mostrar mais
    </button>
  </section>
  ,
</div>
`;

exports[`<TransactionTable /> without transactions history should render <Loading /> 1`] = `
.c7 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c6 {
  background: #000000;
  border: 1px solid #000000;
  border-radius: 4px;
  box-shadow: #DDDDDD 3px 3px 0 0,#000000 3px 3px 0 1px;
  box-sizing: border-box;
  color: #DDDDDD;
  cursor: pointer;
  display: inline-block;
  line-height: 20px;
  overflow: visible;
  padding: 6px 24px;
  touch-action: manipulation;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-user-select: none;
  vertical-align: middle;
  white-space: nowrap;
}

.c6:disabled {
  pointer-events: none;
}

.c6:focus {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c6:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c6:not([disabled]):active {
  background: #DDDDDD;
  outline: 0;
  -webkit-transition: none;
  transition: none;
  box-shadow: #DDDDDD 2px 2px 0 0,#000000 2px 2px 0 1px;
  -webkit-translate: 2px,2px;
  translate: 2px,2px;
}

.c6:has(svg) {
  padding: 0.5% 0.7%;
}

.c6 > svg {
  height: 1.6rem;
  width: 1.6rem;
}

.c5 {
  background: transparent;
  border: none;
  border-bottom: solid 1px #DDDDDD;
  color: #FFFFFF;
  font-size: 1.6rem;
  outline: none;
  width: 10rem;
}

.c5:focus {
  border: none;
  border-bottom: 1px solid #00D775;
  outline: none;
}

.c5 > option {
  background: #000000;
  font-size: 1.6rem;
  height: 0.8rem;
}

.c8:before,
.c8:after {
  border: 0.5rem solid transparent;
  border-radius: 50%;
  content: '';
  left: 50%;
  position: absolute;
  top: 50%;
  -webkit-translate: [-50%,-50%];
  translate: [-50%,-50%];
}

.c8:after {
  -webkit-animation: keBzNL 600ms linear infinite;
  animation: keBzNL 600ms linear infinite;
  border-left: 0.5rem solid #dc143c;
  border-top: 0.5rem solid #dc143c;
  height: 3rem;
  width: 3rem;
}

.c8:before {
  -webkit-animation: keBzNL 1s linear reverse infinite;
  animation: keBzNL 1s linear reverse infinite;
  height: 0.5rem;
  border-left: 0.5rem solid #dc143c;
  border-top: 0.5rem solid #dc143c;
  width: 0.5rem;
}

.c1 {
  font-size: 3.2rem;
  text-transform: uppercase;
}

.c4 {
  font-size: 1.6rem;
  text-transform: none;
}

.c0 {
  text-align: center;
  width: 100%;
}

.c0 > * {
  margin: 2rem 0;
}

.c0 > h1 {
  margin-bottom: 0;
  margin-top: 1rem;
}

.c0 > button {
  margin-top: 0;
}

.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
}

.c3 > p {
  margin: 0;
}

@media (max-width:768px) {
  .c0 {
    width: -webkit-fit-content;
    width: -moz-fit-content;
    width: fit-content;
  }
}

<div>
  <section
    class="c0"
  >
    <h1
      class="c1"
    >
      Histórico de transações
    </h1>
    <section
      class="c2"
    >
      <section
        class="c3"
      >
        <p
          class="c4"
        >
          Tipo:
        </p>
        <select
          class="c5"
        >
          <option
            value="all"
          >
            Todas
          </option>
          <option
            value="cash-in"
          >
            Entradas
          </option>
          <option
            value="cash-out"
          >
            Saídas
          </option>
        </select>
      </section>
      <section
        class="c3"
      >
        <p
          class="c4"
        >
          Ordenar por data:
        </p>
        <button
          class="c6"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="c7"
            fill="currentColor"
            focusable="false"
            name="drop-down"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m7 10 5 5 5-5H7z"
            />
          </svg>
        </button>
      </section>
    </section>
    <div
      class="c8"
    />
  </section>
</div>
`;
